File: webapp/src/lib/auth/jwtExpiry.ts
/** Decode JWT `exp` (seconds) without verifying signature — client-side refresh scheduling only. */
export function getJwtExpMs(accessToken: string): number | null {
  try {
    const parts = accessToken.split('.');
    if (parts.length < 2) return null;
    const payload = JSON.parse(atob(parts[1].replace(/-/g, '+').replace(/_/g, '/'))) as {
      exp?: number;
    };
    if (typeof payload.exp !== 'number') return null;
    return payload.exp * 1000;
  } catch {
    return null;
  }
}

/** Refresh access token this many ms before JWT expiry. */
export const PROACTIVE_REFRESH_LEAD_MS = 5 * 60 * 1000;

/** True when there is no usable access token or it expires within the lead window. */
export function accessTokenNeedsRefresh(accessToken: string | null | undefined): boolean {
  if (!accessToken?.trim()) return true;
  const expMs = getJwtExpMs(accessToken);
  if (!expMs) return true;
  return Date.now() >= expMs - PROACTIVE_REFRESH_LEAD_MS;
}

File: mobile/lib/utils/jwt_expiry.dart
import 'dart:convert';

/// Decode JWT `exp` (seconds) without verifying signature — refresh scheduling only.
int? getJwtExpMs(String accessToken) {
  try {
    final parts = accessToken.split('.');
    if (parts.length < 2) return null;
    var payload = parts[1];
    final mod = payload.length % 4;
    if (mod > 0) payload += '=' * (4 - mod);
    final decoded = utf8.decode(base64Url.decode(payload));
    final map = jsonDecode(decoded) as Map<String, dynamic>;
    final exp = map['exp'];
    if (exp is! num) return null;
    return (exp * 1000).toInt();
  } catch (_) {
    return null;
  }
}

/// Refresh access token this many ms before JWT expiry.
const proactiveRefreshLeadMs = 5 * 60 * 1000;

/// True when there is no usable access token or it expires within the lead window.
bool accessTokenNeedsRefresh(String? accessToken) {
  final token = accessToken?.trim();
  if (token == null || token.isEmpty) return true;
  final expMs = getJwtExpMs(token);
  if (expMs == null) return true;
  return DateTime.now().millisecondsSinceEpoch >= expMs - proactiveRefreshLeadMs;
}
